Ng model options extend

petebacondarwin opened the pull request:
**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**

feat

**What is the current behavior? (You can also link to an open issue here)**

`ngModelOptions` cannot be inherited from ancestor `ngModelOptions`

**What is the new behavior (if this is a feature change)?**

`ngModelOptions` can inherit if specified via `$inherit` values

**Does this PR introduce a breaking change?**

A small internal-ish change to the programmatic API for reading options.

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md#commit-message-format
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

**Other information**:

petebacondarwin committed:
revert: feat(ngModelOptions): allow options to be inherited from ancestor ngModelOptions

This reverts commit 87a2ff76af5d0a9268d8eb84db5755077d27c84c

petebacondarwin committed:
feat(ngModelOptions): allow options to be inherited from ancestor `ngModelOptions`

Previously, you had to apply the complete set of `ngModelOptions` at every point where
you might want to modify just one or two settings.

This change allows more general settings to be applied nearer to the top of the DOM
and then for more specific settings to override those general settings further down
in the DOM.

To prevent unwanted inheritance you must opt-in on a case by case basis:
* To inherit as single property you simply provide the special value `"$inherit"`.
* To inherit all properties not specified locally then include a property `"*": "$inherit"`.

Closes #10922
Closes #15389

BREAKING CHANGE:

The programmatic API for `ngModelOptions` has changed. You must now read options
via the `ngModelController.getOption(name)` method, rather than accessing the option directly as a property
of the `ngModelContoller.$options` object. This does not affect the usage in templates and only
affects custom directives that might have been reading options for their own purposes.

gkalpak commented on test/ng/directive/ngModelOptionsSpec.js:
```diff
@@ -0,0 +1,768 @@
+            '</form>')($rootScope);
+        $rootScope.$digest();
+
+        var inputElm = doc.find('input').eq(0);
```
Nit: Isn't `.eq(0)` redundant?

petebacondarwin replied:
I think that `find` does not return the object that we want (i.e. some collection thing) so we must convert it to get an instance.

petebacondarwin replied:
I see that it is not used elsewhere... I'll try removing it

gkalpak replied:
`eq` returns a new collection with a subset of the items (typically just one):

```js
  eq: function(index) {
      return (index >= 0) ? jqLite(this[index]) : jqLite(this[this.length + index]);
  }
```

Here there is only one `input` element  anyway, so it is redundant.

petebacondarwin commented on src/ng/directive/ngModelOptions.js:
```diff
@@ -0,0 +1,378 @@
+ * @name $defaultModelOptionsProvider
+ * @description
+ *
+ * Here, you can change the default settings from which {@link ngModelOptions}
```
Actually I don't want them to change defaults - ever. That would put us back in the same position as before where application developers can break library components that expect defaults to be a certain value.

Instead if one wants to change stuff globally then they must put an `ngModelOptions` directive on the same element as the `ng-app` attribute. Then use `"*": "$inherit"`

gkalpak replied:
It's fine by me. But then why do we need a public service for that? People will always be able to overwrite that.

petebacondarwin replied:
Good point. I'll change that

petebacondarwin commented on src/ng/directive/ngModelOptions.js:
```diff
@@ -0,0 +1,378 @@
+     * specified.
+     */
+    $get: function() {
+      return new ModelOptions(defaultOptions, defaultOptions);
```
Are you suggesting we create a global variable!?!
I agree.

petebacondarwin commented on src/ng/directive/ngModelOptions.js:
```diff
@@ -0,0 +1,378 @@
+ * @name ngModelOptions
+ *
+ * @description
+ * This directive allows you to modify the behaviour of ngModel and input directives within your
```
OK OK

petebacondarwin commented on src/ng/directive/ngModelOptions.js:
```diff
@@ -0,0 +1,378 @@
+ * should also invoke {@link ngModel.NgModelController `$rollbackViewValue`} on the relevant input field in
+ * order to make sure it is synchronized with the model and that any debounced action is canceled.
+ *
+ * The easiest way to reference the control's {@link ngModel.NgModelController `$rollbackViewValue`}
```
Not sure. This is just cut-n-paste from the previous implementation

petebacondarwin commented on src/ng/directive/ngModelOptions.js:
```diff
@@ -0,0 +1,378 @@
+ * and the {@link $defaultModelOptions} service.
+ */
+function ModelOptions(defaultOptions, options, parentOptions) {
+  this._defaultOptions = defaultOptions;
```
OK

petebacondarwin commented on src/ng/directive/ngModelOptions.js:
```diff
@@ -0,0 +1,378 @@
+        if (key === '*') {
+          inheritAll = true;
+        } else {
+          options[key] = isDefined(parentOptions[key]) ? parentOptions[key] : defaultOptions[key];
```
OK

petebacondarwin commented on test/ng/directive/ngModelOptionsSpec.js:
```diff
@@ -0,0 +1,768 @@
+/* globals generateInputCompilerHelper: false */
+describe('ngModelOptions', function() {
+
+  describe('$defaultModelOptions', function() {
```
Actually they are not :-P
I just need to fix the docs - see my other comment

petebacondarwin commented on test/ng/directive/ngModelOptionsSpec.js:
```diff
@@ -0,0 +1,768 @@
+        expect(inputOptions.getOption('updateOnDefault')).toEqual($defaultModelOptions.getOption('updateOnDefault'));
+        expect(inputOptions.getOption('debounce')).toEqual($defaultModelOptions.getOption('debounce'));
+        expect(inputOptions.getOption('getterSetter')).toEqual($defaultModelOptions.getOption('getterSetter'));
+        expect(inputOptions.getOption('allowInvalid')).toEqual($defaultModelOptions.getOption('allowInvalid'));
```
Right

gkalpak commented on src/ng/directive/ngModelOptions.js:
```diff
@@ -0,0 +1,378 @@
+   * Returns the value of the given option
+   */
+  getOption: function(name) {
+    return isDefined(this._options[name]) ? this._options[name] : this._defaultOptions[name];
```
It wouldn't make any difference anyway. Even with the current implementation we would return `this._defaultOptions[key]` which would be undefined if we didn't specify the option in `$defaultModelOptions`. 

But I agree it is a good idea to have all options in `$defaultModelOptions`.

petebacondarwin commented on src/ng/directive/ngModelOptions.js:
```diff
@@ -0,0 +1,358 @@
+  $$processOptions: function() {
+    var options = this.$$options;
+    // updateOn and updateOnDefault
+    if (isDefined(options.updateOn) && options.updateOn.trim()) {
```
Right! Now this is a good review point.  I knew there must be something that needed fixing. I'll take a look right now.

petebacondarwin committed:
feat(ngModelOptions): allow options to be inherited from ancestor `ngModelOptions`

Previously, you had to apply the complete set of `ngModelOptions` at every point where
you might want to modify just one or two settings.

This change allows more general settings to be applied nearer to the top of the DOM
and then for more specific settings to override those general settings further down
in the DOM.

To prevent unwanted inheritance you must opt-in on a case by case basis:
* To inherit as single property you simply provide the special value `"$inherit"`.
* To inherit all properties not specified locally then include a property `"*": "$inherit"`.

Closes #10922
Closes #15389

BREAKING CHANGE:

The programmatic API for `ngModelOptions` has changed. You must now read options
via the `ngModelController.getOption(name)` method, rather than accessing the option directly as a property
of the `ngModelContoller.$options` object. This does not affect the usage in templates and only
affects custom directives that might have been reading options for their own purposes.

gkalpak commented on src/ng/directive/ngModelOptions.js:
```diff
@@ -0,0 +1,345 @@
+        } else {
+          options[key] = this.$$options[key];
+        }
+      } else if (key === 'updateOn') {
```
There are still some issues with this implementation:

- If `{updateOn: '$inherit'}` it will inherit the `updateOn`, but not the `updateOnDefault` (which might be inherited by `$defaultModelOptions`).

petebacondarwin replied:
Great. I am really working your reviewing skills today!

petebacondarwin committed:
feat(ngModelOptions): allow options to be inherited from ancestor `ngModelOptions`

Previously, you had to apply the complete set of `ngModelOptions` at every point where
you might want to modify just one or two settings.

This change allows more general settings to be applied nearer to the top of the DOM
and then for more specific settings to override those general settings further down
in the DOM.

To prevent unwanted inheritance you must opt-in on a case by case basis:
* To inherit as single property you simply provide the special value `"$inherit"`.
* To inherit all properties not specified locally then include a property `"*": "$inherit"`.

Closes #10922
Closes #15389

BREAKING CHANGE:

The programmatic API for `ngModelOptions` has changed. You must now read options
via the `ngModelController.getOption(name)` method, rather than accessing the option directly as a property
of the `ngModelContoller.$options` object. This does not affect the usage in templates and only
affects custom directives that might have been reading options for their own purposes.

petebacondarwin commented:
OK, trying again. See if you can find another failing scenario @gkalpak

gkalpak commented on test/ng/directive/ngModelOptionsSpec.js:
```diff
@@ -180,6 +180,31 @@ describe('ngModelOptions', function() {
+      it('should `updateOnDefault` as well if we have `updateOn: "$inherit"`', function() {
+        var container = $compile(
+                  '<div ng-model-options="{updateOn: \'keyup\'}">' +
+                    '<input ng-model-options="{\'updateOn\': \'$inherit\'}">' +
```
Unnecessary quotes around `updateOn` (here and below).

gkalpak reviewed (approved):
Fourth time's a charm :smiley:
LGTM
